data-loader/core/src/main/java/com/scalar/db/dataloader/core/dataimport/processor/ImportProcessor.java

```diff
@@ -62,41 +62,77 @@ public abstract class ImportProcessor {
    * @param reader the {@link BufferedReader} used to read the source file
    */
   public void process(int dataChunkSize, int transactionBatchSize, BufferedReader reader) {
-    ExecutorService dataChunkExecutor = Executors.newSingleThreadExecutor();
+    ExecutorService dataChunkReaderExecutor = Executors.newSingleThreadExecutor();
+    ExecutorService dataChunkProcessorExecutor =
+        Executors.newFixedThreadPool(params.getImportOptions().getMaxThreads());
     BlockingQueue<ImportDataChunk> dataChunkQueue =
         new LinkedBlockingQueue<>(params.getImportOptions().getDataChunkQueueSize());
 
+    // Semaphore controls concurrent task submissions, small buffer to be two times of threads
+    Semaphore taskSemaphore = new Semaphore(params.getImportOptions().getMaxThreads() * 2);
```

Comment:
Instead of using a `Semaphore`, I think we can use an `ExecutorService` with a bounded blocking queue, as shown below:

```java
    ExecutorService dataChunkProcessorExecutor =
        new ThreadPoolExecutor(
            params.getImportOptions().getMaxThreads(),
            params.getImportOptions().getMaxThreads(),
            0L,
            TimeUnit.MILLISECONDS,
            new ArrayBlockingQueue<>(params.getImportOptions().getMaxThreads() * 2),
            Executors.defaultThreadFactory(),
            new ThreadPoolExecutor.CallerRunsPolicy());
```

What do you think?